lgn-worker/src/main.rs

```diff
@@ -249,24 +301,37 @@ async fn run(cli: Cli) -> anyhow::Result<()> {
         match inbound.next().await {
             Some(Ok(msg)) => {
                 counter!("zkmr_worker_messages_total").increment(1);
-
-                let result = process_message_from_gateway(
-                    &mut provers_manager,
-                    msg,
-                    &mut outbound,
-                    &mp2_requirement,
-                )
-                .await;
-
-                match result {
-                    Ok(()) => {
-                        counter!("zkmr_worker_messages_successful_total").increment(1);
-                    },
-                    Err(err) => {
-                        counter!("zkmr_worker_messages_error_total").increment(1);
-                        error!("task processing failed. err: {:?}", err);
-                    },
-                }
+                let task_id = msg.task_id.clone();
+
+                let outbound_msg =
+                    match process_message_from_gateway(&mut provers_manager, msg, &mp2_requirement)
+                        .await
+                    {
+                        Ok(serialised_reply) => {
+                            counter!("zkmr_worker_messages_successful_total").increment(1);
+                            WorkerToGwRequest {
+                                request: Some(lagrange::worker_to_gw_request::Request::WorkerDone(
+                                    WorkerDone {
+                                        task_id,
+                                        reply: Some(Reply::TaskOutput(serialised_reply)),
+                                    },
+                                )),
+                            }
+                        },
+                        Err(err) => {
+                            counter!("zkmr_worker_messages_error_total").increment(1);
+                            WorkerToGwRequest {
+                                request: Some(lagrange::worker_to_gw_request::Request::WorkerDone(
+                                    WorkerDone {
+                                        task_id,
+                                        reply: Some(Reply::WorkerError(format!("{:?}", err))),
+                                    },
+                                )),
+                            }
+                        },
+                    };
+
+                outbound.send(outbound_msg).await?;
```

Comment:
The IO was previously split among `run` and `process_message_from_gateway`. Moving the IO to a single place will make it easier to implement retries later on.